packages/xsnap/src/object-inspect.js

```diff
@@ -0,0 +1,489 @@
+}
+function isString(obj) {
+  return (
+    String(obj) === obj && !(typeof obj === 'object' && toStringTag in obj)
+  );
+}
+function isNumber(obj) {
+  return (
+    Number(obj) === obj && !(typeof obj === 'object' && toStringTag in obj)
+  );
+}
+function isBoolean(obj) {
+  return (
+    Boolean(obj) === obj && !(typeof obj === 'object' && toStringTag in obj)
+  );
+}
+
+// Symbol and BigInt do have Symbol.toStringTag by spec, so that can't be used to eliminate false positives
+function isSymbol(obj) {
+  return typeof obj === 'symbol';
+}
+
+function isBigInt(obj) {
+  return typeof obj === 'bigint';
+}
+
+function has(obj, key) {
+  return hasOwn.call(obj, key);
+}
+
+function toStr(obj) {
+  return objectToString.call(obj);
+}
+
+function nameOf(f) {
+  if (f.name) {
+    return f.name;
+  }
+  const m = $match.call(functionToString.call(f), /^function\s*([\w$]+)/);
+  if (m) {
+    return m[1];
+  }
+  return null;
+}
+
+function indexOf(xs, x) {
+  if (xs.indexOf) {
+    return xs.indexOf(x);
+  }
+  for (let i = 0, l = xs.length; i < l; i += 1) {
+    if (xs[i] === x) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+function isMap(x) {
+  if (!mapSize || !x || typeof x !== 'object') {
+    return false;
+  }
+  try {
+    mapSize.call(x);
+    try {
+      setSize.call(x);
+    } catch (s) {
+      return true;
+    }
+    return x instanceof Map; // core-js workaround, pre-v2.5.0
+  } catch (e) {}
+  return false;
+}
+
+function isWeakMap(x) {
+  if (!weakMapHas || !x || typeof x !== 'object') {
+    return false;
+  }
+  try {
+    weakMapHas.call(x, weakMapHas);
+    try {
+      weakSetHas.call(x, weakSetHas);
+    } catch (s) {
+      return true;
+    }
+    return x instanceof WeakMap; // core-js workaround, pre-v2.5.0
+  } catch (e) {}
+  return false;
+}
+
+function isWeakRef(x) {
+  if (!weakRefDeref || !x || typeof x !== 'object') {
+    return false;
+  }
+  try {
+    weakRefDeref.call(x);
+    return true;
+  } catch (e) {}
+  return false;
+}
+
+function isSet(x) {
+  if (!setSize || !x || typeof x !== 'object') {
+    return false;
+  }
+  try {
+    setSize.call(x);
+    try {
+      mapSize.call(x);
+    } catch (m) {
+      return true;
+    }
+    return x instanceof Set; // core-js workaround, pre-v2.5.0
+  } catch (e) {}
+  return false;
+}
+
+function isWeakSet(x) {
+  if (!weakSetHas || !x || typeof x !== 'object') {
+    return false;
+  }
+  try {
+    weakSetHas.call(x, weakSetHas);
+    try {
+      weakMapHas.call(x, weakMapHas);
+    } catch (s) {
+      return true;
+    }
+    return x instanceof WeakSet; // core-js workaround, pre-v2.5.0
+  } catch (e) {}
+  return false;
+}
+
+function inspectString(str, opts) {
+  if (str.length > opts.maxStringLength) {
+    const remaining = str.length - opts.maxStringLength;
+    const trailer = `... ${remaining} more character${
+      remaining > 1 ? 's' : ''
+    }`;
+    return (
+      inspectString($slice.call(str, 0, opts.maxStringLength), opts) + trailer
```

Comment:
Anything I should do about that?

Reply:
We add a trailer which starts with a non surrogate char, so it may just technically be an invalid string, but all implementations handle malformed strings like that.

Reply:
@mhofman I don't see how that trailer helps. In the case in question, it would already be invalid as a unicode string.

Reply:
I think what I meant is that this won't be concatenated with another string which may start with the second half of another surrogate pair, so if we split in the middle, the first half will just be detected as a broken pair, and processed as such by every algorithms which handles strings.

Reply:
Thanks. I understand now, and it is a valid concern.